doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,258 @@
+# Universal, synchronous and in-thread loader hooks
+
+For background and motivation, see https://github.com/nodejs/node/issues/52219. Prototype implementation is in https://github.com/joyeecheung/node/tree/sync-hooks.
+
+TL;DR: the top priority of this proposal is to allow sun-setting CJS loader monkey patching as non-breakingly as possible (which is why it needs to be synchronous and in-thread because that's how the CJS loader and `require()` works). Then it's API consistency with the existing `module.register()` off-thread hooks.
+
+Existing users of CJS loader monkey patching to look into:
+
+- pirates (used by Babel and nyc)
+- require-in-the-middle (used by many tracing agents)
+- yarn pnp
+- tsx
+- ts-node
+
+## API design
+
+A high-level overview of the API.
+
+```js
+const { addHooks, removeHooks } = require('module');
+
+function resolve(specifier, context, nextResolve) {
+  const resolved = nextResolve(specifier, context);
+  return { url: resolved.url.replaceAll(/foo/g, 'bar'); }
+}
+
+function load(url, context, nextLoad) {
+  const loaded = nextLoad(specifier, context);
+  return { source: loaded.source.toString().replaceAll(/foo/g, 'bar'); }
+}
+
+// id is a symbol
+const id = addHooks({ resolve, load, ... });
+removeHooks(id);
+```
+
+1. The names `addHooks` and `removeHooks` take inspiration from pirates. Can be changed to other better names.
+2. An alternative design to remove the hooks could be `addHooks(...).unhook()`, in this case `addHooks()` returns an object that could have other methods.
```

Comment:
The `[Symbol.dispose]()` method on this object, and possibly the friendly `.dispose()` could let this coexist with a lot of the new disposable stuff that's landing.

Reply:
Yeah it would be interesting to use it with the using proposal…

Reply:
I thought about how to compose it but it looks weird.

```
const hook = module.registerHooks(...);
hook.deregister();  // Calls [Symbol.dispose]();

// What happens if we want to allow the hook to be re-added?
```

Reply:
Or we could add `[Symbol.dispose]()` which calls `hook.deregister()` but don't make it part of the direct interface..